operator_rolling_upgrades.md

```diff
@@ -0,0 +1,54 @@
+# Operator Rolling Upgrades
+Optionally allow a Infinispan CR cluster to be upgraded with zero downtime and no loss of data.
+
+## Spec
+Add a field to the CR spec that determines whether an upgrade should be `rolling` or `gracefulShutdown`. The latter is the current approach that does not allow data migration.
+
+```yaml
+apiVersion: infinispan.org/v1
+kind: Infinispan
+metadata:
+  name: example-infinispan
+spec:
+  upgrades: rolling
+```
+
+## Procedure
+When an upgrade is possible, the operator performs the following for each Infinispan CR:
+
+1. Create a new StatefulSet for the target cluster based upon the CR spec, wait until the cluster is formed
+
+2. Retrieve all cache configurations from the cluster in the original Statefulset
+
+3. Append a remote-store configuration to each of the ^ configurations and create cache target cluster:
```

Comment:
Do we support local caches? They will need to be created node by node?

Reply:
That's a good question. Do local caches even make sense in openshift? As the pods are always accessed via a service or route it's not simple to connect to a specific node.

If we did support upgrading local caches, then the process would be something like:
1. Map the source and target pods, so `source-0 -> target-0, `source-1 -> target-1` etc
2. For each pod
  i. Create the local cache with the remote store, with the store pointing to the corresponding source pod
  ii. Sync the local cache via rest
  iii. Remove the remote-store from the configuration